cdap-examples/SpamClassifier/src/main/scala/co/cask/cdap/examples/sparkstreaming/SpamClassifierProgram.scala

```diff
@@ -0,0 +1,117 @@
+
+    val sparkContext: SparkContext = new SparkContext
+
+    LOG.info("Reading stream {} to build classification model", SpamClassifier.STREAM)
+
+    // read the training data stream
+    val trainingData: RDD[String] = sparkContext.fromStream(SpamClassifier.STREAM, 0, Long.MaxValue)
+
+    val termFrequencies = new HashingTF(numFeatures = 1000)
+
+    // spam messages
+    var spam: RDD[String] = trainingData.filter(x => x.startsWith("spam"))
+    spam = spam.map { line =>
+      val strings = line.split("\t")
+      strings(1)
+    }
+
+    // not spam messages
+    var ham: RDD[String] = trainingData.filter(x => x.startsWith("ham"))
+    ham = ham.map { line =>
+      val strings = line.split("\t")
+      strings(1)
+    }
+
+    val spamLabeledPoint: RDD[LabeledPoint] = spam.map(line => termFrequencies.transform(line.split(" "))).map(x =>
+      new LabeledPoint(1, x))
+    val hamLabeledPoint: RDD[LabeledPoint] = ham.map(line => termFrequencies.transform(line.split(" "))).map(x =>
+      new LabeledPoint(0, x))
+
+    val modelData: RDD[LabeledPoint] = spamLabeledPoint.union(hamLabeledPoint)
+    val model: NaiveBayesModel = NaiveBayes.train(modelData, 1.0)
+
+    LOG.info("Built a NaiveBayes model for classification with training data of size: {}", modelData.count())
+
+    val streamingContext: StreamingContext = new StreamingContext(sparkContext, Seconds(2))
+
+    // Create direct kafka stream with brokers and topics
+    val brokers: String = sec.getRuntimeArguments.get("kafka.brokers")
+    require(!Strings.isNullOrEmpty(brokers), "A comma separated list of kafka brokers must be specified. For example:" +
+      " broker1-host:port,broker2-host:port")
+    val topics: String = sec.getRuntimeArguments.get("kafka.topics")
+    require(!Strings.isNullOrEmpty(topics), "A comma separated list of kafka topics must be specified. For example: " +
+      "topic1,topic2")
+
+    val topicsSet = topics.split(",").toSet
+    val kafkaParams = Map[String, String]("metadata.broker.list" -> brokers, "auto.offset.reset" -> "smallest")
+
+    LOG.info("Trying to create a DStream for Kafka with kafka params {} and topics {}", kafkaParams, topicsSet: Any)
+
+    val kafkaData = KafkaUtils.createDirectStream[Array[Byte], String, DefaultDecoder, StringDecoder](streamingContext,
+      kafkaParams, topicsSet)
+
+    // transform the kafka DStream to a key-value pair of message-id and message
+    val messages = kafkaData.map { line: ((Array[Byte], String)) =>
+      val messageSplit: Array[String] = line._2.split(":", 2)
+      (messageSplit(0), messageSplit(1))
+    }
+
+    // for each RDD in the DStream transform it a RDD of message-id and label (spam or ham)
+    messages.foreachRDD { (rdd: RDD[(String, String)]) =>
+      val classified: RDD[(Array[Byte], Double)] = rdd.map { line =>
+        val vector: Vector = termFrequencies.transform(line._2.split(" "))
+        val value: Double = model.predict(vector)
+        (Bytes.toBytes(line._1), value)
```

Comment:
It might make sense to have the time as part of the key as well. You can get the time through the `foreachRDD` method:

```scala
messages.foreachRDD { (rdd, time) =>
  ...
}
```